test/e2e/helpers.go

```diff
@@ -0,0 +1,175 @@
+	return cmd.Run()
+}
+
+// kubectlPatch patches a Kubernetes resource
+func kubectlPatch(resourceType, name, patchType, patch string) error {
+	GinkgoHelper()
+	cmd := exec.Command("kubectl", "patch", resourceType, name,
+		"--type="+patchType, "-p", patch)
+	_, err := utils.Run(cmd)
+	return err
+}
+
+// getControllerPodName retrieves the name of the controller manager pod
+func getControllerPodName(namespace string) (string, error) {
+	GinkgoHelper()
+	jsonpath := "{.items[?(@.metadata.deletionTimestamp==null)].metadata.name}"
+	output, err := kubectlGet("pods", "", namespace, jsonpath)
+	if err != nil {
+		return "", err
+	}
+	return strings.Fields(output)[0], nil
+}
+
+// deleteClusterResources deletes multiple cluster-scoped resources
+func deleteClusterResources(resources []string) {
+	GinkgoHelper()
+	for _, resource := range resources {
+		cmd := exec.Command("kubectl", "delete", resource, "--ignore-not-found")
+		_, _ = utils.Run(cmd)
+	}
+}
+
+// diagnoseAndCleanupStuckTemplates checks for templates stuck with finalizers
+// and performs emergency cleanup if needed for suite teardown.
+// This function waits for controller to process finalizer removal, then diagnoses
+// stuck templates and force-removes finalizers as a last resort.
+func diagnoseAndCleanupStuckTemplates() {
+	GinkgoHelper()
+
+	By("checking for stuck templates with finalizers (emergency cleanup)")
+	// Wait up to 30s for controller to process finalizer removal
+	cmd := exec.Command("kubectl", "wait", "workspacetemplate", "--all", "--all-namespaces",
+		"--for=delete", "--timeout=30s")
+	if err := cmd.Run(); err != nil {
+		// Templates still exist - diagnose before emergency cleanup
+		cmd = exec.Command("kubectl", "get", "workspacetemplate", "-A",
+			"-o", "jsonpath={.items[*].metadata.name}")
+		if output, _ := utils.Run(cmd); len(output) > 0 && strings.TrimSpace(string(output)) != "" {
+			templates := strings.Fields(strings.TrimSpace(string(output)))
+			_, _ = GinkgoWriter.Write([]byte("\n⚠️  WARNING: Templates still exist after workspace cleanup: " + strings.Join(templates, ", ") + "\n"))
+
+			// Diagnose: Check which workspaces reference each stuck template
+			for _, tmplName := range templates {
+				cmd = exec.Command("kubectl", "get", "workspace", "-A",
+					"-o", "jsonpath={.items[?(@.spec.templateRef==\""+tmplName+"\")].metadata.name}")
+				if wsOutput, _ := utils.Run(cmd); len(wsOutput) > 0 && strings.TrimSpace(string(wsOutput)) != "" {
+					_, _ = GinkgoWriter.Write([]byte("   Template '" + tmplName + "' is referenced by workspaces: " +
+						strings.TrimSpace(string(wsOutput)) + " (test leaked resources)\n"))
+				} else {
+					_, _ = GinkgoWriter.Write([]byte("   Template '" + tmplName + "' has NO workspace references - CONTROLLER BUG?\n"))
+				}
+			}
+
+			// Emergency cleanup: Force-remove finalizers to allow suite teardown
+			By("EMERGENCY: Force-removing stuck template finalizers for clean teardown")
```

Comment:
non-blocking: isn't this...hiding a deeper problem w/ template lifecycles?

Reply:
If this emergency cleanup triggers it logs a warning that indicates a controller bug requiring investigation.  Without it, a stuck finalizer would silently timeout during test cleanup, potentially masking the root cause. So this cleanup makes bugs visible, not invisible.

Kubernetes E2E tests have similar finalizer cleanup logic https://github.com/kubernetes/kubernetes/blob/854e67bb51e177b4b9c012928d8271704e9cb80d/test/e2e/apimachinery/namespace.go#L600-L620